File: frontend/src/lib/data-capabilities.ts
/** 数据集能力矩阵的纯函数与配色(2026-09-12, 借鉴 tick-stock-panel 能力路由视图)。
 *
 * 配色约定(对齐"涨跌色只用于价格"): 正常=emerald, 降级=amber, 无可用源=red,
 * 未测量=灰(刻意不与"正常"同色, 避免把没测过看成没问题)。
 */

export type CapabilityStatus = 'ok' | 'degraded' | 'unknown' | 'unavailable'

export interface CapabilitySource {
  provider: string
  name: string | null
  enabled: boolean
  priority: number | null
  success_rate: number | null
  samples: number | null
  /** 读数依据: ewma=进程内滚动窗口(重启归零) / db=累计计数 / none=两者都不足 */
  basis?: 'ewma' | 'db' | 'none'
  ewma_latency_ms: number | null
  last_error: string
}

/** 驱动本次结论的那个源(后端 classify 用的就是它, 与"按 priority 路由生效"是两回事)。 */
export interface CapabilityVerdict {
  provider: string
  success_rate: number
  samples: number
  basis: 'ewma' | 'db' | 'none'
}

export interface CapabilityItem {
  type: string
  label: string
  status: CapabilityStatus
  status_label: string
  reason: string
  sources: CapabilitySource[]
  /** 未测量/无可用源时为 null —— 此时**不许**显示成功率。 */
  verdict: CapabilityVerdict | null
  enabled_count: number
  latest_date: string | null
  age_days: number | null
}

export interface CapabilitySummary {
  total: number
  ok: number
  degraded: number
  unknown: number
  unavailable: number
  degraded_labels: string[]
  unavailable_labels: string[]
}

export interface CapabilitiesResp {
  items: CapabilityItem[]
  summary: CapabilitySummary
  /** 判定门槛(后端 MIN_SAMPLES 透传): 样本不足时 UI 显示进度而不是结论。 */
  min_samples?: number
}

export const STATUS_DOT: Record<CapabilityStatus, string> = {
  ok: 'bg-emerald-500',
  degraded: 'bg-amber-500',
  unknown: 'bg-gray-400',
  unavailable: 'bg-red-500',
}

export const STATUS_TEXT: Record<CapabilityStatus, string> = {
  ok: 'text-emerald-600 dark:text-emerald-400',
  degraded: 'text-amber-600 dark:text-amber-400',
  unknown: 'text-muted-foreground',
  unavailable: 'text-red-600 dark:text-red-400',
}

/** 侧栏一行摘要: "数据能力 12/15"; 有异常时附前两个降级/缺失项。 */
export function capabilitySummary(s: CapabilitySummary | undefined): string {
  if (!s) return ''
  const bad = [...s.unavailable_labels, ...s.degraded_labels]
  const head = `数据能力 ${s.ok}/${s.total}`
  return bad.length ? `${head} · ${bad.slice(0, 2).join('、')}${bad.length > 2 ? ` 等${bad.length}项` : ''}` : head
}

/** 排序: 有问题的排前面(无可用源 → 降级 → 未测量 → 正常), 同级按名称。 */
export function sortItemsByRisk(items: CapabilityItem[]): CapabilityItem[] {
  const rank: Record<CapabilityStatus, number> = { unavailable: 0, degraded: 1, unknown: 2, ok: 3 }
  return [...items].sort((a, b) => rank[a.status] - rank[b.status] || a.label.localeCompare(b.label, 'zh'))
}

/** 生效源 = 启用源里 priority 最小者(与运行期兜底顺序一致); 无启用源返回 null。 */
export function effectiveSource(item: CapabilityItem): CapabilitySource | null {
  const on = item.sources.filter((s) => s.enabled)
  if (!on.length) return null
  return on.reduce((best, s) =>
    (s.priority ?? 99) < (best.priority ?? 99) ? s : best, on[0])
}

/** 矩阵"读数"列(KI-050): 有结论就显示**驱动结论那个源**的成功率; 没结论就显示
 * 样本进度而不是百分比 —— 绝不允许出现"未测量 · 100%"或"正常 · --"这种自相矛盾。 */
export function capabilityRateCell(item: CapabilityItem, minSamples?: number): string {
  if (item.verdict) return `${Math.round(item.verdict.success_rate * 100)}%`
  if (item.status === 'unavailable') return ''
  if (!minSamples) return ''
  const n = Math.max(0, ...item.sources.filter((s) => s.enabled).map((s) => s.samples ?? 0))
  return `样本 ${n}/${minSamples}`
}

File: src/core/data_capabilities.py
"""数据集能力视图(2026-09-12, 借鉴 tick-stock-panel 的能力路由矩阵)。

我们和它的差别: 它按"数据集 × 源能力"做**路由**(可切生效源); 我们的源是运行期按
priority 兜底的, 不需要再做一个路由器。真正缺的是**视图** —— 用户(老板)现在看不出
"哪类数据此刻是好的、哪类在降级、哪类根本没测过", 只能看到一串源名字。

口径:
- 状态四档, **不合并"未测量"到"正常"**(对齐 0.3「降级不伪装」):
  ok 正常 / degraded 降级 / unknown 未测量 / unavailable 无可用源
- degraded 判据: 该数据集所有启用源里, 有足够样本(≥MIN_SAMPLES)且成功率 < OK_SUCCESS_RATE。
- unknown: 有启用源但没有任何可信样本(从未调用/样本太少) → 前端标灰, 不显示绿色。
- 新鲜度只对"我们确实落库"的数据集探测(FRESHNESS_PROBES), 其余给 None —— 不编造。
"""
from __future__ import annotations

# 数据集 → 中文名(与 web/api/datasources.TYPE_LABELS 对齐, 并补齐 DB 里存在但未登记的类型)
DATASET_LABELS = {
    "kline": "K线数据",
    "quote": "实时行情",
    "capital_flow": "资金流向",
    "board_capital_flow": "板块资金",
    "market_capital_flow": "大盘资金",
    "flash_news": "快讯",
    "news": "新闻资讯",
    "events": "事件日历",
    "macro_calendar": "财经日历",
    "fundamentals": "基本面",
    "dragon_tiger": "龙虎榜",
    "margin": "融资融券",
    "shareholders": "股东户数",
    "dividend": "分红",
    "northbound": "北向资金",
    "chart": "K线截图",
    "more_info": "扩展指标",
    "wenda": "通达信问答",
}

# 数据集 → (落库表, 日期列): 只对确实入库的类型探新鲜度
FRESHNESS_PROBES = {
    "kline": ("klines", "ts"),
    "quote": ("quote_snapshots", "ts"),
    "dragon_tiger": ("dragon_tiger_events", "trade_date"),
    "board_capital_flow": ("board_daily", "date"),
}

OK_SUCCESS_RATE = 0.8       # 成功率≥此值算正常
MIN_SAMPLES = 10            # 样本不足不下"降级"结论(避免冷启动误判)

STATUS_LABELS = {"ok": "正常", "degraded": "降级", "unknown": "未测量", "unavailable": "无可用源"}


def _source_view(row: dict, health: dict) -> dict:
    h = health.get(row.get("provider") or "") or {}
    return {
        "provider": row.get("provider"),
        "name": row.get("name"),
        "enabled": bool(row.get("enabled")),
        "priority": row.get("priority"),
        "success_rate": h.get("success_rate"),
        "samples": h.get("count"),
        "ewma_latency_ms": h.get("ewma_latency_ms"),
        "last_error": h.get("last_error") or "",
    }


def classify(sources: list[dict]) -> tuple[str, str]:
    """由该数据集的源列表 + 健康指标定状态。返回 (status, reason)。"""
    enabled = [s for s in sources if s["enabled"]]
    if not enabled:
        return "unavailable", "没有启用的数据源"
    measured = [s for s in enabled if (s["samples"] or 0) >= MIN_SAMPLES and s["success_rate"] is not None]
    if not measured:
        return "unknown", f"{len(enabled)} 个源尚无足够样本(需 ≥{MIN_SAMPLES} 次调用)"
    best = max(measured, key=lambda s: s["success_rate"])
    if best["success_rate"] >= OK_SUCCESS_RATE:
        return "ok", f"最佳源 {best['provider']} 成功率 {round(best['success_rate'] * 100)}%"
    worst = ", ".join(f"{s['provider']} {round(s['success_rate'] * 100)}%" for s in measured)
    return "degraded", f"全部源成功率低于 {round(OK_SUCCESS_RATE * 100)}%({worst})"


def _age_days(latest: str | None, today: str | None) -> int | None:
    """最新数据距今天数(兼容 'YYYY-MM-DD' 与 'YYYYMMDD' 两种日期写法)。"""
    if not latest or not today:
        return None
    from datetime import date

    def _parse(v: str) -> date | None:
        s = str(v)[:10]
        try:
            if "-" in s:
                return date.fromisoformat(s)
            if len(s) >= 8 and s.isdigit():
                return date(int(s[:4]), int(s[4:6]), int(s[6:8]))
        except ValueError:
            return None
        return None

    d1, d2 = _parse(latest), _parse(today)
    return (d2 - d1).days if d1 and d2 else None


def build_capabilities(rows: list[dict], health: dict, freshness: dict | None = None,
                       today: str | None = None) -> dict:
    """rows: DataSource 记录(dict); health: vendor→指标; freshness: 表名→最新日期字符串。"""
    freshness = freshness or {}
    by_type: dict[str, list[dict]] = {}
    for r in rows:
        t = r.get("type") or ""
        if t:
            by_type.setdefault(t, []).append(r)

    items = []
    for t in sorted(by_type):
        sources = sorted((_source_view(r, health) for r in by_type[t]),
                         key=lambda s: (not s["enabled"], s["priority"] if s["priority"] is not None else 99))
        status, reason = classify(sources)
        table, _col = FRESHNESS_PROBES.get(t, (None, None))
        latest = str(freshness[table]) if (table and freshness.get(table) is not None) else None
        items.append({
            "type": t,
            "label": DATASET_LABELS.get(t, t),
            "status": status,
            "status_label": STATUS_LABELS[status],
            "reason": reason,
            "sources": sources,
            "enabled_count": sum(1 for s in sources if s["enabled"]),
            "latest_date": latest,
            "age_days": _age_days(latest, today),
        })

    def _count(st):
        return sum(1 for i in items if i["status"] == st)

    return {
        "items": items,
        "summary": {
            "total": len(items),
            "ok": _count("ok"),
            "degraded": _count("degraded"),
            "unknown": _count("unknown"),
            "unavailable": _count("unavailable"),
            "degraded_labels": [i["label"] for i in items if i["status"] == "degraded"],
            "unavailable_labels": [i["label"] for i in items if i["status"] == "unavailable"],
        },
    }
